File: apps/web/src/lib/env-validator.ts
/**
 * Environment variable validation module for Next.js frontend.
 * Validates required environment variables at build/runtime.
 */

export class EnvValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'EnvValidationError';
  }
}

interface ValidationRule {
  key: string;
  required: boolean;
  allowedValues?: string[];
  default?: string;
}

/**
 * Validate a single environment variable.
 */
function validateEnvVar(rule: ValidationRule): string | undefined {
  const { key, required, allowedValues, default: defaultValue } = rule;
  const value = process.env[key] || defaultValue;

  if (required && !value) {
    throw new EnvValidationError(`Required environment variable '${key}' is not set`);
  }

  if (allowedValues && value && !allowedValues.includes(value)) {
    throw new EnvValidationError(
      `Environment variable '${key}' has invalid value '${value}'. ` +
      `Allowed values: ${allowedValues.join(', ')}`
    );
  }

  return value;
}

/**
 * Validate all required environment variables.
 * Call this at the top of your app or in a root layout.
 */
export function validateAllEnvVars(): Record<string, string | undefined> {
  const validated: Record<string, string | undefined> = {};
  const errors: string[] = [];

  const rules: ValidationRule[] = [
    {
      key: 'NEXT_PUBLIC_FIREBASE_API_KEY',
      required: true,
    },
    {
      key: 'NEXT_PUBLIC_FIREBASE_AUTH_DOMAIN',
      required: true,
    },
    {
      key: 'NEXT_PUBLIC_FIREBASE_PROJECT_ID',
      required: true,
    },
    {
      key: 'NEXT_PUBLIC_FIREBASE_STORAGE_BUCKET',
      required: true,
    },
    {
      key: 'NEXT_PUBLIC_FIREBASE_MESSAGING_SENDER_ID',
      required: true,
    },
    {
      key: 'NEXT_PUBLIC_FIREBASE_APP_ID',
      required: true,
    },
    {
      key: 'NEXT_PUBLIC_API_BASE',
      required: false, // Optional for relative URL usage
    },
    {
      key: 'NEXT_PUBLIC_DIRECT_API_URL',
      required: false, // Optional, has default
    },
    {
      key: 'NEXT_PUBLIC_ALLOWED_DIRECT_API_DOMAINS',
      required: false, // Optional, has default
    },
    {
      key: 'NODE_ENV',
      required: false,
      allowedValues: ['development', 'production', 'test'],
      default: 'development',
    },
  ];

  for (const rule of rules) {
    try {
      const value = validateEnvVar(rule);
      validated[rule.key] = value;
    } catch (error) {
      if (error instanceof EnvValidationError) {
        errors.push(error.message);
      } else {
        throw error;
      }
    }
  }

  if (errors.length > 0) {
    const errorMessage = 'Environment variable validation failed:\n' +
      errors.map(err => `  - ${err}`).join('\n');
    throw new EnvValidationError(errorMessage);
  }

  if (process.env.NODE_ENV === 'development') {
    console.log('✅ Environment variables validated successfully');
  }

  return validated;
}

/**
 * Validate environment variables and exit if validation fails (for server-side).
 */
export function validateEnvVarsOrExit(): void {
  try {
    validateAllEnvVars();
  } catch (error) {
    if (error instanceof EnvValidationError) {
      console.error(`❌ Startup failed due to environment validation errors:\n${error.message}`);
      process.exit(1);
    }
    throw error;
  }
}

File: services/agent-api/app/env_validator.py
"""
Environment variable validation module.
Validates required environment variables at application startup.
"""

import os
import sys
from typing import Dict, List, Optional
import logging

logger = logging.getLogger(__name__)


class EnvValidationError(Exception):
    """Raised when environment variable validation fails."""
    pass


def _get_bool(key: str, default: str = "false") -> bool:
    """Helper to get boolean environment variable."""
    return os.getenv(key, default).lower() == "true"


def validate_env_var(
    key: str,
    required: bool = False,
    allowed_values: Optional[List[str]] = None,
    default: Optional[str] = None
) -> Optional[str]:
    """
    Validate a single environment variable.

    Args:
        key: Environment variable key
        required: Whether the variable is required
        allowed_values: List of allowed values (if applicable)
        default: Default value if not set

    Returns:
        The environment variable value, or default if not set

    Raises:
        EnvValidationError: If validation fails
    """
    value = os.getenv(key, default)

    if required and not value:
        raise EnvValidationError(f"Required environment variable '{key}' is not set")

    if allowed_values and value and value not in allowed_values:
        raise EnvValidationError(
            f"Environment variable '{key}' has invalid value '{value}'. "
            f"Allowed values: {', '.join(allowed_values)}"
        )

    return value


def validate_all_env_vars() -> Dict[str, str]:
    """
    Validate all required environment variables at startup.

    Returns:
        Dict of validated environment variables

    Raises:
        EnvValidationError: If any validation fails
    """
    validated = {}
    errors = []

    # Environment type validation
    try:
        environment = validate_env_var(
            "ENVIRONMENT",
            required=False,
            allowed_values=["development", "staging", "production"],
            default="development"
        )
        validated["ENVIRONMENT"] = environment
    except EnvValidationError as e:
        errors.append(str(e))
        environment = "development"  # Fallback for later checks

    # Firebase configuration
    try:
        firebase_project_id = validate_env_var(
            "FIREBASE_PROJECT_ID",
            required=True
        )
        validated["FIREBASE_PROJECT_ID"] = firebase_project_id
    except EnvValidationError as e:
        errors.append(str(e))

    try:
        firebase_storage_bucket = validate_env_var(
            "FIREBASE_STORAGE_BUCKET",
            required=True
        )
        validated["FIREBASE_STORAGE_BUCKET"] = firebase_storage_bucket
    except EnvValidationError as e:
        errors.append(str(e))

    # Google Cloud configuration
    use_vertexai = _get_bool("GOOGLE_GENAI_USE_VERTEXAI", "false")
    if use_vertexai:
        try:
            google_cloud_project = validate_env_var(
                "GOOGLE_CLOUD_PROJECT",
                required=True
            )
            validated["GOOGLE_CLOUD_PROJECT"] = google_cloud_project
        except EnvValidationError as e:
            errors.append(str(e))

        try:
            google_cloud_location = validate_env_var(
                "GOOGLE_CLOUD_LOCATION",
                required=False,
                default="global"
            )
            validated["GOOGLE_CLOUD_LOCATION"] = google_cloud_location
        except EnvValidationError as e:
            errors.append(str(e))
    else:
        # If not using Vertex AI, validate GOOGLE_API_KEY instead
        try:
            google_api_key = validate_env_var(
                "GOOGLE_API_KEY",
                required=True
            )
            validated["GOOGLE_API_KEY"] = google_api_key
        except EnvValidationError as e:
            errors.append(str(e))

    # Gemini models
    try:
        gemini_model = validate_env_var(
            "GEMINI_MODEL",
            required=False,
            default="gemini-2.5-flash"
        )
        validated["GEMINI_MODEL"] = gemini_model
    except EnvValidationError as e:
        errors.append(str(e))

    # CORS configuration (required in production)
    if environment == "production":
        try:
            allowed_origins = validate_env_var(
                "ALLOWED_ORIGINS",
                required=True
            )
            validated["ALLOWED_ORIGINS"] = allowed_origins
        except EnvValidationError as e:
            errors.append(str(e))

        # Sentry DSN (recommended in production)
        try:
            sentry_dsn = validate_env_var("SENTRY_DSN")
            if not sentry_dsn:
                logger.warning(
                    "SENTRY_DSN is not set in production environment. "
                    "Error monitoring will not be available."
                )
            else:
                validated["SENTRY_DSN"] = sentry_dsn
        except EnvValidationError as e:
            errors.append(str(e))

    if errors:
        error_message = "Environment variable validation failed:\n" + "\n".join(f"  - {err}" for err in errors)
        logger.error(error_message)
        raise EnvValidationError(error_message)

    logger.info(f"Environment variables validated successfully for environment: {environment}")
    return validated


def validate_env_vars_on_startup() -> None:
    """
    Validate environment variables on application startup.
    Exits the application if validation fails.
    """
    try:
        validate_all_env_vars()
    except EnvValidationError as e:
        logger.critical(f"Startup failed due to environment validation errors:\n{e}")
        sys.exit(1)
